File: project/jimmer-sql/src/test/java/org/babyfish/jimmer/sql/model/embedded/Machine.java
package org.babyfish.jimmer.sql.model.embedded;

import org.babyfish.jimmer.Formula;
import org.babyfish.jimmer.sql.*;
import org.jetbrains.annotations.Nullable;

import java.util.*;

@Entity
public interface Machine {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    long id();

    /**
     * Embeddable property
     */
    @Key
    Location location();

    @Nullable
    @PropOverride(prop = "host", columnName = "SECONDARY_HOST")
    @PropOverride(prop = "port", columnName = "SECONDARY_PORT")
    Location secondaryLocation();

    @Formula(dependencies = {"location.host", "secondaryLocation.host"})
    default List<String> hosts() {
        List<String> hosts = new ArrayList<>();
        hosts.add(location().host());
        Location secondaryLocation = secondaryLocation();
        if (secondaryLocation != null) {
            hosts.add(secondaryLocation.host());
        }
        return hosts;
    }

    int cpuFrequency();

    int memorySize();

    int diskSize();

    MachineDetail detail();

    @Formula(dependencies = {"detail.factories"})
    default int factoryCount() {
        return detail().factories().size();
    }

    @Formula(dependencies = {"detail.factories"})
    default Set<String> factoryNames() {
        Map<String, String> factories = detail().factories();
        return new LinkedHashSet<>(factories.keySet());
    }
}

File: project/jimmer-sql-kotlin-test-model/src/main/kotlin/org/babyfish/jimmer/sql/kt/model/embedded/Machine.kt
package org.babyfish.jimmer.sql.kt.model.embedded

import org.babyfish.jimmer.Formula
import org.babyfish.jimmer.sql.*

@Entity
interface Machine {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    val id: Long

    /**
     * Embedded property
     */
    val location: Location

    val detail: MachineDetail

    @Formula(dependencies = ["detail.factories"])
    val factoryCount: Int
        get() = detail.factories.size

    @Formula(dependencies = ["detail.factories"])
    val factoryNames: Set<String>
        get() = detail.factories.keys.toSet()
}

